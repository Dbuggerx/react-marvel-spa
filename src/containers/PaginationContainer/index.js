// @flow

import React, { Component } from 'react';
import Pagination from '../../components/Pagination';

type Props = {
    offset: number,
    limit: number,
    total: number,
    goToPage: (page: number) => void,
};

export default class PaginationContainer extends Component<Props> {
    get pageCount() {
        return Math.ceil(this.props.total / this.props.limit);
    }

    get currentPage() {
        return Math.ceil(this.pageCount - ((this.props.total - this.props.offset) / this.props.limit));
    }

    goToPreviousPage = () => {
        if (this.currentPage === 1) return;
        this.props.goToPage(this.currentPage - 1);
    };

    goToNextPage = () => {
        if (this.currentPage === this.pageCount) return;
        this.props.goToPage(this.currentPage + 1);
    };

    render() {
        return (
            <Pagination
                pageCount={this.pageCount}
                currentPage={this.currentPage}
                goToNextPage={this.goToNextPage}
                goToPreviousPage={this.goToPreviousPage}
            />
        );
    }
}
