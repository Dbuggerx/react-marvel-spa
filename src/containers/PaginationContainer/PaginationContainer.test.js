import React from 'react';
import { shallow } from 'enzyme';
import PaginationContainer from './index';

describe('PaginationContainer', () => {
    describe('render', () => {
        test('renders Pagination', () => {
            const props = {
                offset: 3,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('pageCount', () => {
        test('returns correct value', () => {
            const props = {
                offset: 3,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            expect(wrapper.instance().pageCount).toBe(10);
        });
    });

    describe('currentPage', () => {
        test('returns correct value', () => {
            const props = {
                offset: 30,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            expect(wrapper.instance().currentPage).toBe(3);
        });
    });

    describe('goToPreviousPage', () => {
        test('calls "props.goToPage"', () => {
            const props = {
                offset: 30,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            wrapper.instance().goToPreviousPage();
            expect(props.goToPage).toHaveBeenCalledWith(2);
        });

        test('doesn\'t call "props.goToPage" when already on the first page', () => {
            const props = {
                offset: 3,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            wrapper.instance().goToPreviousPage();
            expect(props.goToPage).not.toHaveBeenCalled();
        });
    });

    describe('goToNextPage', () => {
        test('calls "props.goToPage"', () => {
            const props = {
                offset: 30,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            wrapper.instance().goToNextPage();
            expect(props.goToPage).toHaveBeenCalledWith(4);
        });

        test('doesn\'t call "props.goToPage" when already on the last page', () => {
            const props = {
                offset: 100,
                limit: 10,
                total: 100,
                goToPage: jest.fn(),
            };
            const wrapper = shallow(<PaginationContainer {...props} />);
            wrapper.instance().goToNextPage();
            expect(props.goToPage).not.toHaveBeenCalled();
        });
    });
});
