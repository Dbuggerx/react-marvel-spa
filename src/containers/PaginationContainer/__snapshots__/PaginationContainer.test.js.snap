// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationContainer render renders Pagination 1`] = `
<Pagination
  currentPage={1}
  goToNextPage={[Function]}
  goToPreviousPage={[Function]}
  pageCount={10}
/>
`;
